elastic/apm-server
# Copy Java attacher jar to a tmp directory

## Motivation/summary

Fixing a bug discovered while testing this end-to-end: when Elastic Agent is ran through `./elastic-agent install`, it creates a file tree that is not readable for all users on Linux by default. 
Therefore, when running the attacher as a non-root user, it can't access the attacher jar.
The fix is to create a tmp directory with write permissions only to root and copy the attacher jar to it with `0755` permissions to make it accessible for all users for read and execute.

Fixe verified on:
- Amazon Linux 2
- Ubuntu 22.04

Note for reviewer:
- pay extra care to security concerns I may have overlooked
- currently, the agent is downloaded and verified for each attachment, but we should consider ways to prevent that
- we currently assume the default temp dir would work, but in some OSs we may need to have a tmp dir per user
- the new test may fail on Windows, where file modes are handled differently, we'll take care of that later

## Checklist

- [ ] Documentation has been updated - would be handled separately through #8708 

## How to manually test these changes

1. setup a stack and install a matching agent on the tested Linux host
2. set Elastic agent log level to `debug`
3. run a Java app on the tested host **as a limited-privileges user.** It is recommended that this app will have a way to identify it specifically through matching rules, for example, use `-Delastic.apm.attach=true` in the command line that runs it
4. within the Fleet UI, edit the corresponding agent's policy to include the APM integration. In the APM integration, within the APM agent tab, enable the Java agent auto-attachment option and make sure the matching rule includes a hint that will specifically identify your Java app process
5. apply the policy and inspect the logs. It should contain indications that attachment took place. In addition, it should contain the path of the temp dir to which the attacher was copied
6. login to the Linux host and look for the tmp dir indicated by logs - it should have `0700` mode and be owned by the same user that runs the Java app. It should contain the attacher jar with `0600` mode and be owned by the same user. If attachment was successful, it should contain an `agent` folder with internal structure that includes some PGP-utility libraries and a directory per agent version as requested within the integration.
7. verify that APM data is being traced and available in Kibana
8. restart the Java app, invoke some traffic on it and verify that the same agent jar was used and that APM data is available 
9. stop Elastic agent and verify that the attacher temp dir is removed

## Related issues

#8718 
#8660 
#6969 
#7832
